lib/resque/plugins/retry.rb

```diff
@@ -452,6 +464,86 @@ def clean_retry_key(*args)
         log_message 'clean_retry_key', args
         Resque.redis.del(redis_retry_key(*args))
       end
+
+      # Returns the try again callbacks.
+      #
+      # @return [Array<Proc>]
+      #
+      # @api public
+      def try_again_callbacks
+        @try_again_callbacks ||= []
+      end
+
+      # Register a try again callback that will be called when the job fails
+      # but is trying again.
+      #
+      # @example Using a try again callback
+      #
+      #   try_again_callback do |exception, *args|
+      #     logger.error(
+      #       "Resque job received exception #{exception} and is trying again")
+      #   end
+      #
+      # @yield [exception, *args]
+      # @yieldparam exception [Exception] the exception that was raised
+      # @yieldparam args [Array] job arguments
+      #
+      # @api public
+      def try_again_callback(&block)
+        try_again_callbacks << block
+      end
+
+      # Runs all the try again callbacks.
+      #
+      # @param exception [Exception]
+      # @param args [Object...]
+      #
+      # @api private
+      def run_try_again_callbacks(exception, *args)
+        try_again_callbacks.each do |callback|
+          instance_exec(exception, *args, &callback)
```

Comment:
That's a good question... Well we don't catch exceptions from the retry criteria checks either, right?

I see some scenarios for an exception:
1. We catch the exception for them, and provide some way for the user to deal with it -- this is just like letting them do the begin/rescue themselves.
2. We catch the exception for them but do nothing with it -- this seems bad.
3. We don't catch the exception -- so the code is fine as is.
4. We catch the exception, and trigger the normal "retry" logic based on that Exception as usual. Which... in practice seems to be similar to not catching the exception (if the job isn't supposed to retry on that exception) or catching the exception and ignoring it (if the job is supposed to retry on that exception). This also sounds a bit more complicated to implement too.

Anyway, I feel like we should not catch the exception, and include a warning in the README.